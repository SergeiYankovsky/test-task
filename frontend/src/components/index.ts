export * from "./Input";
export * from "./Button";
export * from "./Card";
export * from "./Header";
export * from "./Modal";
export * from "./Form";
