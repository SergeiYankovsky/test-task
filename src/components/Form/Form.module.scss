.form {
  max-width: 536px;
  width: 100%;
  border-radius: 8px;
  background-color: #ffffff;
}
