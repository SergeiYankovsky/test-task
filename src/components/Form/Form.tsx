import { FC, PropsWithChildren } from "react";
import styles from "./Form.module.scss";

export interface FormProps extends PropsWithChildren {
  onSubmit: () => void;
}
export const Form: FC<FormProps> = ({ onSubmit, children }) => {
  return (
    <form className={styles.form} onSubmit={onSubmit}>
      {children}
    </form>
  );
};
