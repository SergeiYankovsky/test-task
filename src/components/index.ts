export * from "./Input";
export * from "./Button";
export * from "./Form";
export * from "./Card";
export * from "./Header";
export * from "./Modal";
