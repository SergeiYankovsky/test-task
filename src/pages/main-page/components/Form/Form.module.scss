.data_container {
  padding: 10px 20px;
}

.title {
  font-size: 24px;
  color: #707070;
}

.input_container {
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  column-gap: 16px;
  margin: 32px 0;

  & > :nth-child(3) {
    grid-column: span 2;
  }
}

.button_container {
  display: flex;
  padding: 8px 20px;
  gap: 8px;
  background-color: #ebecec;
  border-bottom-left-radius: 8px;
  border-bottom-right-radius: 8px;
}
.right_section_buttons {
  display: flex;
  gap: 8px;
  margin-left: auto;
}

@media screen and (max-width: 500px) {
  .input_container {
    display: flex;
    flex-direction: column;
  }
}
