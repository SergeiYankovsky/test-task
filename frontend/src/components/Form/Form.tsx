import { FC } from "react";
import styles from "./Form.module.scss";
import { Contact } from "@/types";
import { SubmitHandler, useForm } from "react-hook-form";
import { Button, Input } from "@/components";

type FormProps = {
  contact: Partial<Contact> | null;
  onCloseModal: () => void;
  onAddUser: (contact: Contact) => void;
  onUpdateUser: (contact: Contact) => void;
  onDeleteUser: (id: number) => void;
};

export const Form: FC<FormProps> = ({
  contact,
  onCloseModal,
  onAddUser,
  onUpdateUser,
  onDeleteUser,
}) => {
  const {
    register,
    handleSubmit,
    formState: { isValid },
  } = useForm<Contact>({
    defaultValues: { ...contact },
  });

  const onSubmit: SubmitHandler<Contact> = (contact) => {
    if (!contact.id) {
      onAddUser(contact);
      onCloseModal();
      return;
    }
    onUpdateUser(contact);
    onCloseModal();
  };

  const handleDelete = (id?: number) => {
    if (id) {
      onDeleteUser(id);
      onCloseModal();
    }
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)} data-testid="form" className={styles.form}>
      <div className={styles.data_container}>
        <h2 className={styles.title}>Eintrag bearbeiten</h2>
        <div className={styles.input_container}>
          <Input
            data-testid="VornameInput"
            placeholder="Vorname*"
            {...register("firstname", { required: true })}
          />
          <Input
            data-testid="NachnameInput"
            placeholder="Nachname*"
            {...register("lastname", { required: true })}
          />
          <Input
            data-testid="E-MailInput"
            placeholder="E-Mail*"
            {...register("email", {
              required: true,
              pattern: {
                value: /\S+@\S+\.\S+/,
                message: "Entered value does not match email format",
              },
            })}
          />
        </div>
      </div>
      <div className={styles.button_container}>
        {contact?.id && (
          <Button variant="secondary" onClick={() => handleDelete(contact.id)}>
            LÃ–SCHEN
          </Button>
        )}
        <div className={styles.right_section_buttons}>
          <Button variant="primary" onClick={onCloseModal}>
            ABBRECHEN
          </Button>
          <Button variant="default" disabled={!isValid}>
            SPEICHERN
          </Button>
        </div>
      </div>
    </form>
  );
};
