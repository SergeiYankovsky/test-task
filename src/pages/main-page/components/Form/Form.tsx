import { FC } from "react";
import styles from "./Form.module.scss";
import { Contact } from "../../../../types/types";
import { SubmitHandler, useForm } from "react-hook-form";
import { Button, Input, Form as FormUI } from "../../../../components";

type FormProps = {
  contact: Partial<Contact> | null;
  onAddUser: (contact: Contact) => void;
  onUpdateUser: (contact: Contact) => void;
  onDeleteUser: (id: number) => void;
  onCloseModal: () => void;
};

export const Form: FC<FormProps> = ({
  contact,
  onAddUser,
  onDeleteUser,
  onUpdateUser,
  onCloseModal,
}) => {
  const {
    register,
    handleSubmit,
    formState: { isValid },
  } = useForm<Contact>({
    defaultValues: { ...contact },
  });
  const onSubmit: SubmitHandler<Contact> = (contact) => {
    if (!contact.id) {
      onAddUser(contact);
      onCloseModal();
      return;
    }
    onUpdateUser(contact);
    onCloseModal();
  };

  const handleDelete = () => {
    if (!contact?.id) return;
    onDeleteUser(contact.id);
    onCloseModal();
  };
  return (
    <FormUI onSubmit={handleSubmit(onSubmit)} data-testid="form">
      <div className={styles.data_container}>
        <h2 className={styles.title}>Eintrag bearbeiten</h2>
        <div className={styles.input_container}>
          <Input
            data-testid="VornameInput"
            placeholder="Vorname*"
            {...register("firstname", { required: true })}
          />
          <Input
            data-testid="NachnameInput"
            placeholder="Nachname*"
            {...register("lastname", { required: true })}
          />
          <Input
            data-testid="E-MailInput"
            placeholder="E-Mail*"
            {...register("email", {
              required: true,
              pattern: {
                value: /\S+@\S+\.\S+/,
                message: "Entered value does not match email format",
              },
            })}
          />
        </div>
      </div>
      <div className={styles.button_container}>
        {contact?.id && (
          <Button type="secondary" onClick={handleDelete}>
            LÃ–SCHEN
          </Button>
        )}
        <div className={styles.right_section_buttons}>
          <Button type="primary" onClick={onCloseModal}>
            ABBRECHEN
          </Button>
          <Button type="default" disabled={!isValid}>
            SPEICHERN
          </Button>
        </div>
      </div>
    </FormUI>
  );
};
