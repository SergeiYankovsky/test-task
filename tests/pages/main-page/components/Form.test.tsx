import { Contact } from "../../../../src/types/types";
import { Form } from "../../../../src/pages/main-page/components/Form";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Form
        contact={null}
        onAddUser={function (contact: Contact): void {
          throw new Error("Function not implemented.");
        }}
        onUpdateUser={function (contact: Contact): void {
          throw new Error("Function not implemented.");
        }}
        onDeleteUser={function (id: number): void {
          throw new Error("Function not implemented.");
        }}
        onCloseModal={function (): void {
          throw new Error("Function not implemented.");
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
