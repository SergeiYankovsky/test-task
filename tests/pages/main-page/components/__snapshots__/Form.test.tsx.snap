// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="data_container"
  >
    <h2
      className="title"
    >
      Eintrag bearbeiten
    </h2>
    <div
      className="input_container"
    >
      <div
        className="container"
      >
        <input
          className="input"
          name="firstname"
          onChange={[Function]}
          placeholder="Vorname*"
        />
        <label
          className="label"
        >
          Vorname*
        </label>
      </div>
      <div
        className="container"
      >
        <input
          className="input"
          name="lastname"
          onChange={[Function]}
          placeholder="Nachname*"
        />
        <label
          className="label"
        >
          Nachname*
        </label>
      </div>
      <div
        className="container"
      >
        <input
          className="input"
          name="email"
          onChange={[Function]}
          placeholder="E-Mail*"
        />
        <label
          className="label"
        >
          E-Mail*
        </label>
      </div>
    </div>
  </div>
  <div
    className="button_container"
  >
    <div
      className="right_section_buttons"
    >
      <button
        className="button button_primary"
        onClick={[Function]}
      >
        ABBRECHEN
      </button>
      <button
        className="button button_default"
        disabled={true}
      >
        SPEICHERN
      </button>
    </div>
  </div>
</form>
`;
